fixes cnn grad failures

quickwritereader opened the pull request:
## What changes were proposed in this pull request?

    

1.  enable testing. add missing seeds for some configurations.

2.  restricts maxpool2d_bp Cudnn allowing only EWS cases. it will use our Cuda implementation for non-EWScases. Here are brief details:
    - Cudnn API  call was giving zeros for non-ews gradO input which were the reason for the gradient failures
    - Cudnn API is poorly documented, and it needs 4 different arguments indeed. passed the first two arrays either are the same (pooling2dBpCUDNN) or calculated with additional code. see $cudnnUtils.cu#L230-L240 vs cudnnUtils.cu#L380-L404
    - it is decided to handle non-EWS cases by our implementation instead of writing a workaround for the Cudnn API call

## How was this patch tested?

unit tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [x] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [x] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [x] Created tests for any significant new code additions.
- [x] Relevant tests for your changes are passing.

agibsonccc commented on deeplearning4j/deeplearning4j-core/src/test/java/org/deeplearning4j/gradientcheck/CNN1DGradientCheckTest.java:
```diff
@@ -105,7 +104,7 @@ void testCnn1DWithLocallyConnected1D() {
                         }
                     }
-                    MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder().dataType(DataType.DOUBLE).updater(new NoOp()).dist(new NormalDistribution(0, 1)).convolutionMode(ConvolutionMode.Same).list().layer(new Convolution1DLayer.Builder().activation(afn).kernelSize(kernel).stride(stride).padding(padding).nIn(convNIn).nOut(convNOut1).rnnDataFormat(RNNFormat.NCW).build()).layer(new Lo ...
+                    MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder().seed(2021).dataType(DataType.DOUBLE).updater(new NoOp()).dist(new NormalDistribution(0, 1)).convolutionMode(ConvolutionMode.Same).list().layer(new Convolution1DLayer.Builder().activation(afn).kernelSize(kernel).stride(stride).padding(padding).nIn(convNIn).nOut(convNOut1).rnnDataFormat(RNNFormat.NCW).build()).l ...
```
Mind adding some comments about how important the seed is here? A log of issues we ran in to with certain tests is handy.

quickwritereader replied:
refactored all calls into one method and added the comment

agibsonccc commented on libnd4j/include/ops/declarable/platform/cudnn/maxpool2d.cu:
```diff
@@ -95,6 +95,16 @@ PLATFORM_IMPL(maxpool2d_bp, ENGINE_CUDA) {
     const auto isNCHW = block.getIArguments()->size() > 10 ? !INT_ARG(10) : 1;         // INT_ARG(10): 0-NCHW, 1-NHWC
 
+#if 0
+    input->printShapeInfo("2d bp input");
```
Mind removing the debugging code?

agibsonccc commented on libnd4j/include/ops/declarable/platform/cudnn/maxpool2d.cu:
```diff
@@ -123,9 +136,12 @@ PLATFORM_CHECK(maxpool2d_bp, ENGINE_CUDA) {
-    return goodType && (input->dataType() == gradO->dataType())
-                    && (input->dataType() == gradI->dataType())
-                    && shape::haveSameShapeAndStrides(input->shapeInfo(), gradI->shapeInfo());
+    return  goodType && (input->dataType() == gradO->dataType())
```
Mind adding a minor comment paragraph explaining the issues found here so we know why we're defaulting to our implementation in certain cases?

quickwritereader replied:
a short comment was added. but more detailed information is written in the commit message. it's usable in Github blames

quickwritereader committed:
enable CNN1DGradientCheckTest. add missing seed for the NeuralNetConfiguration instance.

Signed-off-by: AbdelRauf <rauf@konduit.ai>

quickwritereader committed:
restricts maxpool2d_bp Cudnn allowing only EWS cases. it will use our Cuda implementation for non-EWScases. Here are brief details:
    - Cudnn API call was giving zeros for non-ews gradO input which were the reason for the gradient failures
    - Cudnn API needs 4 different arguments indeed. in the current code base passed the first two arrays either are the same (pooling2dBpCUDNN) or calculated with additional code. see $cudnnUtils.cu#L230-L240 vs cudnnUtils.cu#L380-L404
    - it is decided to handle non-EWS cases by our implementation instead of writing a workaround for the Cudnn API call

Signed-off-by: AbdelRauf <rauf@konduit.ai>
